http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/KubernetesServiceDbQueries.scala

```diff
@@ -0,0 +1,256 @@
+import cats.implicits._
+import org.broadinstitute.dsde.workbench.leonardo.db.nodepoolQuery.unmarshalNodepool
+
+object KubernetesServiceDbQueries {
+  def listFullApps(googleProject: Option[GoogleProject],
+                   labelFilter: LabelMap = Map(),
+                   includeDeleted: Boolean = false)(implicit ec: ExecutionContext): DBIO[List[KubernetesCluster]] =
+    joinFullAppAndUnmarshal(
+      listByProject(googleProject, includeDeleted),
+      includeDeleted match {
+        case true  => nodepoolQuery
+        case false => nodepoolQuery.filter(_.destroyedDate === dummyDate)
+      },
+      includeDeleted match {
+        case true  => appQuery
+        case false => appQuery.filter(_.destroyedDate === dummyDate)
+      },
+      labelFilter
+    )
+
+  //this is intended to be called first by any kubernetes /app endpoints to enforce one cluster per project
+  //an error is thrown if the cluster is creating, due to the fact that we do not allow apps to be created while the cluster for that project is creating, since that means another app is already queued and waiting on this cluster
+  //if the cluster already exists, this is a no-op
+  def saveOrGetForApp(
+    saveKubernetesCluster: SaveKubernetesCluster
+  )(implicit ec: ExecutionContext): DBIO[SaveClusterResult] =
+    for {
+      clusterOpt <- kubernetesClusterQuery.getMinimalActiveClusterByName(saveKubernetesCluster.googleProject)
+
+      failure = DBIO.failed(
+        KubernetesAppCreationException(
+          s"You cannot create an app while a cluster is in ${KubernetesClusterStatus.Precreating} or ${KubernetesClusterStatus.Provisioning}"
+        )
+      )
+      eitherClusterOrError <- clusterOpt match {
+        case Some(cluster) =>
+          cluster.status match {
+            case KubernetesClusterStatus.Precreating  => failure
+            case KubernetesClusterStatus.Provisioning => failure
+            case _                                    => DBIO.successful(SaveClusterResult(cluster, true))
+          }
+        case _ => kubernetesClusterQuery.save(saveKubernetesCluster).map(c => SaveClusterResult(c, false))
+      }
+    } yield eitherClusterOrError
+
+  def getActiveFullAppByName(googleProject: GoogleProject, appName: AppName, labelFilter: LabelMap = Map())(
+    implicit ec: ExecutionContext
+  ): DBIO[Option[GetAppResult]] =
+    joinFullAppAndUnmarshal(listByProject(Some(googleProject)),
+                            nodepoolQuery.filter(_.destroyedDate === dummyDate),
+                            appQuery.findActiveByNameQuery(appName),
+                            labelFilter)
+      .map(clusters =>
+        clusters.headOption.fold[Option[GetAppResult]](
+          None
+        )(cluster =>
+          Some(
+            GetAppResult(
+              //we are guaranteed exactly one app and nodepool here if the query doesn't return None due to the way the query JOINs and database unique enforcement
```

Comment:
This solution took some iteration to arrive on. From the call-site, it is highly desirable not to need to traverse the cluster-> nodepool -> app structure (a cluster contains a list of nodepools, and each nodepool has a list of apps) when we expect one 'app' to be returned.

Reply:
Done. Its a bit ugly, but I wanted the assertions to follow the `DBIO.failure` pattern.